src/engine/MultiColumnJoin.cpp

```diff
@@ -0,0 +1,471 @@
+// Copyright 2018, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Florian Kramer (florian.kramer@netpun.uni-freiburg.de)
+
+#include "./MultiColumnJoin.h"
+
+using std::string;
+
+// _____________________________________________________________________________
+MultiColumnJoin::MultiColumnJoin(QueryExecutionContext* qec,
+                                 std::shared_ptr<QueryExecutionTree> t1,
+                                 std::shared_ptr<QueryExecutionTree> t2,
+                                 const vector<array<Id, 2>>& jcs)
+    : Operation(qec), _joinColumns(jcs), _multiplicitiesComputed(false) {
+  // Make sure subtrees are ordered so that identical queries can be identified.
+  AD_CHECK_GT(jcs.size(), 0);
+  if (t1->asString() < t2->asString()) {
+    _left = t1;
+    _right = t2;
+  } else {
+    _left = t2;
+    _right = t1;
+    for (unsigned int i = 0; i < _joinColumns.size(); i++) {
+      std::swap(_joinColumns[i][0], _joinColumns[i][1]);
+    }
+  }
+}
+
+// _____________________________________________________________________________
+string MultiColumnJoin::asString(size_t indent) const {
+  std::ostringstream os;
+  for (size_t i = 0; i < indent; ++i) {
+    os << " ";
+  }
+  os << "MULTI_COLUMN_JOIN\n" << _left->asString(indent) << " ";
+  os << "join-columns: [";
+  for (size_t i = 0; i < _joinColumns.size(); i++) {
+    os << _joinColumns[i][0] << (i < _joinColumns.size() - 1 ? " & " : "");
+  };
+  os << "]\n";
+  for (size_t i = 0; i < indent; ++i) {
+    os << " ";
+  }
+  os << "|X|\n" << _right->asString(indent) << " ";
+  os << "join-columns: [";
+  for (size_t i = 0; i < _joinColumns.size(); i++) {
+    os << _joinColumns[i][1] << (i < _joinColumns.size() - 1 ? " & " : "");
+  };
+  os << "]";
+  return os.str();
+}
+
+// Used to generate all up to 125 combinations of left, right and result size.
+template <size_t I, size_t J, size_t K>
+struct MultiColumnJoinCaller {
+  static void call(size_t i, size_t j, size_t k,
+                   shared_ptr<const ResultTable> leftResult,
+                   shared_ptr<const ResultTable> rightResult,
+                   const std::vector<std::array<Id, 2>>& joinColumns,
+                   ResultTable* result, size_t resultSize) {
+    if (I == i) {
```

Comment:
Does it make sense to make this `constexpre if`? As far as I can see it's indeed only relevant at compile time. Also the `i, j, k` should probably be `const` as they aren't changed anymore.

Reply:
As i is not a compile time constant we can't make this constexpr. I added the const to the size_t parameters.